tensorflow/python/ops/math_ops.py

```diff
@@ -1788,18 +1788,24 @@ def _AddNShape(op):
 def _SelectShape(op):
```

Comment:
I am assuming you meant math_ops.cc:810.  The Shape function in C++ seems much more complex than necessary. Are those SetShapeFn tested somewhere in C++ ?